### PR TITLE
Generalize the recipient finding logic and reduce the boilerplate code.

### PR DESCRIPTION
**Pull request by Ferroin**
Based on discussion in #3956.

This generalizes the recipient finding logic in alarm-notify.sh by converting it to utilize computed variable names in a loop instead of having it all serialized.  It also handles declaration of the `SEND_*`, `DEFAULT_RECIPIENT_*`, and `role_recipients_*` variables in a similar manner.

Overall, this significantly reduces the amount of boilerplate code needed to add a new notification method.  By just adding the variable tag (in lowercase) to the `method_names` variable near the top, you get automatic declarations of the above mentioned variables and are automatically hooked into the recipient finding logic.

There are two oddities that this has had to work around:

* Kafka:  The kafka notification support does not do anything with recipients.  I know nothing about Kafka myself, so I'm just leaving this alone.  Because it doesn't want the whole recipient handling, it's explicitly left out of the `method_names` variable, and the `SEND_KAFKA` variable is explicitly declared.

* EMail:  The email handling has two peculiarities to it:
    - There is a default value for `DEFAULT_RECIPIENT_EMAIL` explicitly defined in alarm-notify.sh.  We just declare the variable in the normal loop, and then assign it down with the other email related config variables.
    - Recipient names need to be separated by a comma and a space, not just a space like everything else wants.  This is achieved by some creative usage of the `cut` command and the `$IFS` variable. Ideally, this wouldn't be needed, but spaces in email addresses are valid (as stupid as that fact is), so we have to account for that.

There are a couple of places where we need uppercase versions of the method names.  This is currently done using `awk` in a way that should work in any POSIX compliant shell.  If we are willing to restrict this to only supporting Bash 4 or newer, I can rewrite those to avoid the need to fork another process.

This has been tested using alarm-test.sh on my systems.  It appears to not change any behavior.  Further testing from others would probably be a good idea however, as this is poking at somewhat complex stuff in alarm-notify.sh, and if it doesn't work, it will completely stop notifications from being delivered.

**NOTE**: This breaks all outstanding PR's which add notification methods.

The following PR's are affected directly, either because they use code this is changing, or should be using code that this is changing:  #3956 #3932 #3398 #3962 
The following PR's will require either a merge or rebase, but are otherwise not affected directly: #2976

### CODE REVIEW

File: plugins.d/alarm-notify.sh [hunk 1]
```diff
@@ -433,268 +391,43 @@ filter_recipient_by_criticality() {
...
+    declare -A arr=()
+
+    for x in ${roles//,/ } ; do
+        # the roles 'silent' and 'disabled' mean:
+        # don't send a notification for this role
+        [ "${x}" = "silent" -o "${x}" = "disabled" ] && continue
+
+        role_recipients="role_recipients_${method_name}"
+        role_recipients=${!role_recipients}
+        default_recipient="DEFAULT_RECIPIENT_$(echo ${method_name} | awk '{print toupper($0)}')"
```
Comment (ktsaou) [inline plugins.d/alarm-notify.sh]:
```sh
default_recipient="DEFAULT_RECIPIENT_${method_name^^}"
```
does the same

File: plugins.d/alarm-notify.sh [hunk 2]
```diff
@@ -433,268 +391,43 @@ filter_recipient_by_criticality() {
...
+
+        a="${role_recipients[${x}]}"
+        [ -z "${a}" ] && a="${!default_recipient}"
+        for r in ${a//,/ } ; do
+            [ "${r}" != "disabled" ] && filter_recipient_by_criticality email "${r}" && arr[${r/|*/}]="1"
```
Comment (ktsaou) [inline plugins.d/alarm-notify.sh]: `email` should be `${method}`

File: plugins.d/alarm-notify.sh [hunk 3]
```diff
@@ -433,268 +391,43 @@ filter_recipient_by_criticality() {
...
-# build the list of custom recipients
-to_custom="${!arr_custom[*]}"
-[ -z "${to_custom}" ] && SEND_CUSTOM="NO"
+    # build the list of recipients
+    # This needs some creativity because email recipients need to be
+    # separated differently from how the others are.
+    oldifs="$IFS"
+    IFS='	'
+    if [ ${method_name} = "email" ] ; then
+        declare to_${method_name}="$(echo ${!arr[*]} | cut -d '	' --output-delimiter=', ' -f 1-)"
```
Comment (ktsaou) [inline plugins.d/alarm-notify.sh]: I think you should attempt to use the bash way for this, to avoid forks, pipes, etc.

File: plugins.d/alarm-notify.sh [hunk 4]
```diff
@@ -433,268 +391,43 @@ filter_recipient_by_criticality() {
...
+    else
+        declare to_${method_name}="$(echo ${!arr[*]} | cut -d '	' --output-delimiter=' ' -f 1-)"
```
Comment (ktsaou) [inline plugins.d/alarm-notify.sh]: this too

File: plugins.d/alarm-notify.sh [hunk 5]
```diff
@@ -433,268 +391,43 @@ filter_recipient_by_criticality() {
...
+    fi
+    IFS="${oldifs}"
+    send="SEND_$(echo ${method_name} | awk '{print toupper($0)}')"
```
Comment (ktsaou) [inline plugins.d/alarm-notify.sh]: like above, use bash

File: plugins.d/alarm-notify.sh [hunk 6]
```diff
@@ -142,6 +142,12 @@ docurl() {
     return $?
 }
 
+# -----------------------------------------------------------------------------
+# List of all the notification mechanisms we support.
+# Used in a couple of places to write more compact code.
+
+method_names="email pushover pushbullet telegram slack alerta flock discord hipchart twilio messagebird pd fleep syslog custom msteam"
```
Comment (ktsaou) [inline plugins.d/alarm-notify.sh]: I suggest to turn this to multi-line with a backslash at the end of each line, so that we can easily track what is added, removed, changed.

Comment (Ferroin) [reply plugins.d/alarm-notify.sh]: Agreed, except we don't need backslashes.  Bash supports multi-line strings without backslashes, and a newline will work just fine as a separator for separating the items.

File: plugins.d/alarm-notify.sh (hunk 2 above)
Comment (Ferroin) [reply plugins.d/alarm-notify.sh]: Doh!  You're right.  I only tested this with email (because that's the only delivery mechanism I currently use), so it just happened to work.

File: plugins.d/alarm-notify.sh (hunk 1 above)
Comment (Ferroin) [reply plugins.d/alarm-notify.sh]: That also only works in Bash 4.  Looking again at the rest of the script though, we're already doing a number of things that only work in Bash 4, so I'll just update things to do this.

File: plugins.d/alarm-notify.sh (hunk 3 above)
Comment (Ferroin) [reply plugins.d/alarm-notify.sh]: Unless I'm mistaken, there's no way to do this in bash without calling an external command.  `cut` happens to be slightly more efficient in this case than `tr` here.  In theory though, we don't need the call for anything but email (because everything else is just iterating over the list instead of using it directly, and tabs work to separate items for that).

**Comment by Ferroin** [comment]
In addition to addressing the comments, I've pushed a commit that relocates the loop and skips it for delivery methods that are disabled either by the user or because the config or command requirements for them are not met.  This should give a rather nice boost to performance for most users, as the recipient finding code is some of the most computationally expensive code in this script.

File: plugins.d/alarm-notify.sh (hunk 3 above)
Comment (ktsaou) [reply plugins.d/alarm-notify.sh]: What you want to do with this?

in general you can use the following:

convert the associative array keys to a simple array:

```sh
tmp=(  "${!arr[@]}" )
```

Then use `tmp` to get the array members you need:

```sh
start=1
len=${#tmp[@]}
members=$(( len - start ))
methods=( "${tmp[@]:$start:$members}" )
```

So, `${methods[@]}` will have the members you picked.

Comment (Ferroin) [reply plugins.d/alarm-notify.sh]: I could also rewrite it to use the same method that's already being used for email addresses.  I was just trying to avoid yet another loop here (it would end up being a conditional loop, but still), while still accounting for spaces being valid in email addresses (tabs technically are too, but nobody uses them at all).  I'm going to do some quick testing to see if it's actually more efficient done with the loop or done with `cut` (the fork may be a deciding factor, but we're forking all over the place anyway in here).

Comment (Ferroin) [reply plugins.d/alarm-notify.sh]: Hmm, kind of surprises me a bit, but the loop solution currently used in `alarm-notify.sh` for this seems to be the fastest until you get well above 100 recipient addresses (which is not likely to be a common case, and will probably be rejected by many email servers).  I'll update things to use that then instead of calling out to a subshell and cut.

**Comment by ktsaou** [comment]
Is this well tested?
Merging it will affect users significantly, so we need to be sure it will work.

**Comment by Ferroin** [comment]
I've been running it without issues on all my systems for about 10 days now (constituting a total of about a hundred delivered alarm notifications), so I'm reasonably confident it works correctly for how I have alarms configured (one email target, and one syslog target (now that I finally got the syslog delivery set up on all my systems)).  I've also tested it a couple of times using the `alarm-test.sh` script with Pushbullet, Discord, Telegram, and the new Amazon SNS notification methods, and all of them appear to work correctly as well.

That said, it would be good to get some further testing from other people.

Also, on a slight side note, based on some really simple testing with the `time` command and `alarm-test.sh`, it looks like this actually shaves a bit of time off of the delivery process, most likely because it's skipping the recipient finding code for methods that aren't in use.  Exact savings appears to be very dependent on the performance of the system it's being tested on (it's a tiny fraction of a second per-message on my nice Ryzen-based home server, but almost two thirds of a second per-message for my VPS nodes).

**Comment by ktsaou** [comment]
@l2isbad @paulfantom could you please test one?
The changes are straight forward, but this will affect all notifications sent by netdata, so if you can test it I would feel a lot more comfortable to merge it.

**Comment by Ferroin** [comment]
Rebased to handle conflict from the recent config relocation.

File: plugins.d/alarm-notify.sh.in [hunk 7]
```diff
@@ -859,6 +554,51 @@ if [ "${SEND_AWSSNS}" = "YES" -a -z "${aws}" ]
...
+        done
+    done
+
+    # build the list of recipients
+    to="to_${method_name}"
+    if [ ${method_name} = "email" ] ; then
+        declare to_${method_name}
+        for x in "${!arr[@]}" ; do
+            [ ! -z "${!to}" ] && to_email="${!to}, "
+            to_email="${!to}${x}"
```
Comment (ktsaou) [inline plugins.d/alarm-notify.sh.in]: `to_email` ?

**Comment by ktsaou** [comment]
@Ferroin is it possible to unit test this little code?
I mean we could supply a few example configurations and expect specific results from it. Can you check it?

**Comment by Ferroin** [comment]
Yeah, it should be pretty easy actually.  Most logical option to me seems to be adding a mode that bails after calculating recipients and spits out the variable contents.

File: plugins.d/alarm-notify.sh.in (hunk 7 above)
Comment (Ferroin) [reply plugins.d/alarm-notify.sh.in]: That's part of the special-case for email handling.  We ask for email addresses separated by a comma and a space, not just a space, so we need to parse the recipient list differently.

Somewhat ironically, the line right above the first context line shown here is the one that indicates this.

**Comment by Ferroin** [comment]
@ktsaou I just pushed a commit to add the ability to unit test the recipient finding logic and the status filtering logic.  It doesn't include any actual tests, but it should make testing this much easier.

Invocation for unit testing with this patch looks like this:

    alarm-notify.sh unittest <role> <cfg> <status> <old_status>

Where `<role>` is the role to use for the test, `<cfg>` is the configuration file to use, and `<status> and <old_status>` are the simulated statuses to test against.

This will print one line like the following for each method enabled in the config:

    results: <method>: <recipients>

Where `<method>` is the method the recipients were parsed for, and `<recipients>` is a list of the recipients.

File: plugins.d/alarm-notify.sh.in (hunk 7 above)
Comment (ktsaou) [reply plugins.d/alarm-notify.sh.in]: hm... then it should be probably better to run the loop without that `if` and after the loop do:

```sh
# remove multiple adjacent spaces from ${to_email}
to_email="${to_email//  / }"
# remove leading space
to_email="${to_email/# /}"
# remove trailing space
to_email="${to_email/% /}"

# emails have to be separated by comma and a space
to_email="${to_email// /, }"
```

or even (the following is probably better compared to the previous one - it handles multi-space separated values better):

```sh
fix_to_email() {
  to_email=
  while [ ! -z "${1}" ]
  do
    [ ! -z "${to_email}" ] && to_email="${to_email}, "
    to_email="${to_email}${1}"
    shift 1
  done
}

# ${to_email} without quotes here
fix_to_email ${to_email}
```

**Comment by ktsaou** [comment]
This is nice!
Could you please provide also a script to test a few cases?
We could ship this script in the `tests/` directory.

**Comment by ktsaou** [comment]
Also, to avoid confusion, I would also suggest to name all variables that are used like this `${!var}`, with `_var`. So that we will be able to distinguish the easily.

File: plugins.d/alarm-notify.sh.in (hunk 7 above)
Comment (Ferroin) [reply plugins.d/alarm-notify.sh.in]: Or, alternatively, we could push that logic into the function that actually sends the email and just make the API for send functions more consistent, though I'm not quite sure exactly how hard that may be (the diff might be huge, because it might require relocating the template email into the function body).

**Comment by ktsaou** [comment]
So this:

```
send="SEND_${method_name^^}"
```

should be:

```
send_var="SEND_${method_name^^}"
```

File: plugins.d/alarm-notify.sh.in (hunk 7 above)
Comment (ktsaou) [reply plugins.d/alarm-notify.sh.in]: yes, I like that too.
Though I think email is a bit out of scope at the end of the file...

**Comment by Ferroin** [comment]
I should have a commit for separating out the `$to_email` handling within a couple of hours (it's just about lunchtime here, so I'm just going to do it after lunch).  The test cases will probably be some time Monday, as I want to have a chance to make sure anything I come up with actually works before pushing those out, and try to make sure I've got the edge cases covered properly.

**Comment by ktsaou** [comment]
Perfect! You rock!

**Comment by Ferroin** [comment]
I'm thinking something may have gotten messed up in one of the recent rebases, because it looks like this isn't working properly anymore.

**Comment by Ferroin** [comment]
NVM, it was 24a4510 that broke it.  I'll look into getting that fixed.  Independent of that being broken, it looks like the unit test mode does not work completely correctly right now either, but until I've got the recipient finding logic working, that's kind of secondary.

**Comment by ktsaou** [comment]
perfect! looking forward for you test cases...

**Comment by Ferroin** [comment]
Sorry about the delay getting test cases out.  I've been having issues producing stripped down configurations that actually work despite tests done with simple modified versions of the default `health_alarm_notify.conf` file working just fine.

**Comment by Ferroin** [comment]
So far, the testing I've done has been with the stock `health_alarm_notify.conf` file for the config, with minor modifications to add recipients to various delivery mechanisms.  I've been trying to put together minimal test configurations along the lines of:

```
SEND_EMAIL="YES"
role_recipients_email[test]="foo@example.com bar@example.com"
```

But for some reason they don't seem to be working correctly.

**Comment by ktsaou** [comment]
@Ferroin 	do you believe you can finish testing this?

Ideally, I would like to have this merged, and immediately after split it into modules, so that each plugin appears on its own directory. We could a method similar to `charts.d.plugin`, that is require 2 functions:

1. `XXX_check()`, to check user configuration and decide if the plugin is enabled or not.
2. `XXX_send()`, to send the notification.

We could cache the result of `XXX_check()` into a file in `/var/cache/netdata` and re-run the checks for all the modules, if the timestamp of the config file is newer than the timestamp of the cache file. So, if someone enables only `email`, then `alarm-notify.sh` will not source the code of the rest of the plugins at all.

Anyway, I think the first action should be to have this merged. Then we can proceed to the next steps.

**Comment by Ferroin** [comment]
Sorry I hadn't gotten back to this yet.  I have done some more testing on my end, but haven't gotten anything as far as unit tests done yet since I"m still having issues with it not working with minimal configs.

I would sill like to see some positive results from other people testing this, but other than that I think it's ready to merge.

**Comment by ktsaou** [comment]
> I would sill like to see some positive results from other people testing this, but other than that I think it's ready to merge.

If you can't properly test it with unit testing, I would suggest not to merge it. People will not get alarms from netdata, which is a big problem...

**Comment by ktsaou** [comment]
So, you are confident this works?
Merge it?

**Comment by Ferroin** [comment]
I'm reasonably confident that it works correctly right now, but IMO it's your call whether to merge it or not.  Regardless, I'll continue working on getting proper unit testing for this.

**Comment by cakrit** [comment]
@Ferroin please rebase to include PR #4765, which has a few changes in function send_slack.

**Comment by cakrit** [comment]
I tested this with the slack integration and it worked.

**Comment by ktsaou** [comment]
@Ferroin 	I approved this!
You rock!

Just pay some attention now to user issues.

**Comment by Ferroin** [comment]
@ktsaou Will do.

I'll also get started working on getting this properly modularized.
